src/pandas/types/numeric.h

```diff
@@ -44,14 +106,153 @@ class PANDAS_EXPORT IntegerArray : public NumericArray<TYPE> {
+  // this function will be faster than calling GetNullCount() > 0.
+  bool HasNulls() const {
+    for (auto ii = 0; ii < this->length_; ++ii) {
+      if (IsNull(ii)) { return true; }
+    }
+    return false;
+  }
+
+  // Marks the value at the specified index as null. A pre-condition
+  // is that the valid_bits for this object have already been
+  // initialized.
+  void SetNull(int64_t i) {
+    // Pre-condition is that valid_bits_ is non-null because
+    // operations that would set them (e.g. - operator+=)
+    // would first create the valid_bits_ if needed
+    BitUtil::SetBit(
+        static_cast<MutableBuffer&>(*valid_bits_).mutable_data(), this->offset_ + i);
+  }
+
+  // Marks the value at the specified index as null. A pre-condition
+  // is that the valid_bits for this object have already been
+  // initialized.
+  void SetValid(int64_t i) {
+    // Pre-condition is that valid_bits_ is non-null because
+    // operations that would set them (e.g. - operator+=)
+    // would first create the valid_bits_ if needed
+    BitUtil::ClearBit(
+        static_cast<MutableBuffer&>(*valid_bits_).mutable_data(), this->offset_ + i);
+  }
+
+  // Returns whether or not the value at the specified index has
+  // been marked as null.
+  bool IsNull(int64_t i) const {
+    if (valid_bits_) {
+      // TODO: range check?
+      return BitUtil::GetBit(valid_bits_->data(), this->offset_ + i);
+    } else {
+      return false;
+    }
+  }
+
+  // Copy the list of nulls (aka valid bits) to a buffer
+  Status CopyNulls(int64_t length, std::shared_ptr<Buffer>* out) const {
+    // Pre-condition is that we already have valid_bits_ set
+    return valid_bits_->Copy(this->offset_, BitUtil::BytesForBits(length), out);
+  }
+
+  template <typename OTHER_TYPE>
+  IntegerArray<TYPE>& operator+=(const IntegerArray<OTHER_TYPE>& other) {
+    // TODO: check Status and throw exception
+    EnsureMutable();
+    auto length = std::min(this->length_, other.length());
+    if (HasNulls()) {
+      // Perform a |= on the valid bits if the other array has them.
+      // TODO: do this more efficiently, perhaps by passing our
+      //       valid_bits_ like this: other.BitwiseOr(*valid_bits_).
+      //       It might also be useful to have a class that wraps
+      //       much of the valid bit logic.
+      if (other.HasNulls()) {
+        for (auto ii = 0; ii < length; ++ii) {
+          if (IsNull(ii) || other.IsNull(ii)) {
+            SetNull(ii);
+          } else {
+            SetValid(ii);
+          }
+        }
+      }
+    } else if (other.HasNulls()) {
+      std::shared_ptr<Buffer> new_valid_bits;
+      // TODO: check Status and throw an exception
+      other.CopyNulls(length, &new_valid_bits);
+    }
+    auto this_data = this->mutable_data();
+    auto other_data = other.data();
+    if (HasNulls()) {
+      for (auto ii = 0; ii < length; ++ii, ++this_data, ++other_data) {
+        if (!IsNull(ii)) { *this_data += *other_data; }
```

Comment:
I would say to do the arithmetic always and use byte/word-wise `&` with the respective bitmaps (if they are byte-aligned, otherwise do it bitwise) to propagate nulls separately